jsonpath: add `like_regex`, string and null scalars

normanchenn opened the pull request:
#### jsonpath/parser: add support for string scalars

This commit adds string scalars, enabling string comparisons within
jsonpath queries.

Epic: None
Release note (sql change): Support string comparisons within jsonpath
queries.

#### jsonpath/parser: add support for null scalars.

This commit adds null scalars, enabling null comparisons within jsonpath
queries.

Epic: None
Release note (sql change): Support null comparisons within jsonpath
queries.

#### jsonpath: add `like_regex` support

This commit add `like_regex` predicate evaluation support. Flags for
`like_regex` are not supported yet.

Informs: #143243
Epic: None
Release note (sql change): Add `like_regex` predicate evaluation support
for jsonpath queries. Flags for `like_regex` are not supported yet.

normanchenn committed:
jsonpath/parser: add support for string scalars

This commit adds string scalars, enabling string comparisons within
jsonpath queries.

Epic: None
Release note (sql change): Support string comparisons within jsonpath
queries.

normanchenn committed:
jsonpath/parser: add support for null scalars.

This commit adds null scalars, enabling null comparisons within jsonpath
queries.

Epic: None
Release note (sql change): Support null comparisons within jsonpath
queries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143240)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Very nice!

> Reviewed 4 of 4 files at r1, 3 of 3 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-:-OM8owZI4h43ScMGhGKg:b-j7qk9t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`-- commits` line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8ocV-2yvlsBCqFcAL:-OM8ocV-2yvlsBCqFcAM:b4on9ah):*
nit: should we mention that `flag` part is not supported? Or do you think we'll add that soon?
___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8l8ex6f8q4w-ho6-B:-OM8l8ex6f8q4w-ho6-C:b-gq4yqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5350a8293c56cfb68401b86bda7c9983821d7f99/pkg/util/jsonpath/parser/jsonpath.y#L459)):*
> ```Yacc
> 
> unreserved_keyword:
>   AND
> ```

Hm, why AND, NOT, and OR are being removed in the first commit?
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 794 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8lmG93LZAlOPzc3QT:-OM8lmG93LZAlOPzc3QU:b-q06juj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae78464277b7e078122db2efb179b225b8ef49f/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L794)):*
> ```
> 
> query T
> SELECT jsonb_path_query('{}', 'null == null');
> ```

nit: did you mean to make this `!=`?
___
*[`pkg/util/jsonpath/eval/operation.go` line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8oKqg-agE_YbXWHEo:-OM8oKqg-agE_YbXWHEp:b-l6cxr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/86603a37344e3949fd8a664f603e2e8be8be3d46/pkg/util/jsonpath/eval/operation.go#L112)):*
> ```Go
> 		return jsonpathBoolUnknown, err
> 	}
> 	res := r.Find([]byte(*text))
> ```

nit: I think `r.MatchString` would be cleaner here.
___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8lNXh4UTb1Vl5hHQF:-OM8lNXh4UTb1Vl5hHQG:bzbdlkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bf6b6e499b5ab527ec48006eb789d49adf4c2d/pkg/util/jsonpath/parser/jsonpath.y#L460)):*
> ```Yacc
> 
> any_identifier:
>   IDENT
> ```

nit: should we remove IDENT since I think it's no longer used?
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 851 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8oT7l-6wKIBz6Jp9N:-OM8oT7l-6wKIBz6Jp9O:boy41u) ([raw file](https://github.com/cockroachdb/cockroach/blob/86603a37344e3949fd8a664f603e2e8be8be3d46/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L851)):*
> ```
> "987_650"
> "987_424"
> ```

nit: should we add a couple test cases where regex uses a backslash?


<!-- Sent from Reviewable.io -->

normanchenn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143240)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn and @yuzefovich)

___
*[`-- commits` line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8ocV-2yvlsBCqFcAL:-OMEIk0-89d-UULs6CgX:b-lzr50h):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we mention that `flag` part is not supported? Or do you think we'll add that soon?
</blockquote></details>

I'm not sure if we should support it yet - I'm not sure if theres a straightforward way to add these flags with the current implementation.
___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8l8ex6f8q4w-ho6-B:-OMCbIRc5VXlsPeh7Hhs:b-83wv50) ([raw file](https://github.com/cockroachdb/cockroach/blob/5350a8293c56cfb68401b86bda7c9983821d7f99/pkg/util/jsonpath/parser/jsonpath.y#L459)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, why AND, NOT, and OR are being removed in the first commit?
</blockquote></details>

I don't think AND, NOT, and OR should be considered unreserved keywords since when we're scanning for these tokens, they are symbols such as `&&`, `||`, and `!`, so when `lexbase.GetKeywordID` is called, I think it would make more sense to identify the string as an IDENT instead of AND for a key such as "and"
___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8lNXh4UTb1Vl5hHQF:-OMCaG5mEZ0inUom6XSV:bf71ffs) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bf6b6e499b5ab527ec48006eb789d49adf4c2d/pkg/util/jsonpath/parser/jsonpath.y#L460)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we remove IDENT since I think it's no longer used?
</blockquote></details>

We are still using IDENT, because most key accessor names are identified as IDENT during scanning. For example, in `$.a`, a is an IDENT. See the call to `lexbase.GetKeywordID` in `JSONPathScanner.scanIdent`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice! Glad that `UnescapePattern` does what we need (and, philosophically speaking, it makes sense that we'd need to do something very similar for `LIKE` comparison operator). I do have one more comment.

> Reviewed 10 of 10 files at r4, 13 of 13 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-:-OMF5Ye_C6k0nBh_PCFa:b4710oh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8l8ex6f8q4w-ho6-B:-OMF0cHd5vs1lQPH9Ay0:b-gyoicq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5350a8293c56cfb68401b86bda7c9983821d7f99/pkg/util/jsonpath/parser/jsonpath.y#L459)):*
<details><summary><i>Previously, normanchenn (Norman Chen) wrote…</i></summary><blockquote>

I don't think AND, NOT, and OR should be considered unreserved keywords since when we're scanning for these tokens, they are symbols such as `&&`, `||`, and `!`, so when `lexbase.GetKeywordID` is called, I think it would make more sense to identify the string as an IDENT instead of AND for a key such as "and"
</blockquote></details>

I see, sgtm.
___
*[`pkg/util/jsonpath/parser/jsonpath.y` line 460 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8lNXh4UTb1Vl5hHQF:-OMF0Sxh-lOb-1mH9NoZ:b-fm2koa) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bf6b6e499b5ab527ec48006eb789d49adf4c2d/pkg/util/jsonpath/parser/jsonpath.y#L460)):*
<details><summary><i>Previously, normanchenn (Norman Chen) wrote…</i></summary><blockquote>

We are still using IDENT, because most key accessor names are identified as IDENT during scanning. For example, in `$.a`, a is an IDENT. See the call to `lexbase.GetKeywordID` in `JSONPathScanner.scanIdent`.
</blockquote></details>

Ack, thanks, I missed this usage.
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 794 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OM8lmG93LZAlOPzc3QT:-OMF2PH18JBiTycwK4Ja:b2cr63e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ae78464277b7e078122db2efb179b225b8ef49f/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L794)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: did you mean to make this `!=`?
</blockquote></details>

nit: this fixup belongs in the second commit, not third :smiley: 
___
*[`pkg/sql/logictest/testdata/logic_test/jsonb_path_query` line 879 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OMF2n3z46I3EnAfKMfE:-OMF2n3z46I3EnAfKMfF:biamj3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3366e206d87cdb840bc9f2326c4bf396bfac0ed4/pkg/sql/logictest/testdata/logic_test/jsonb_path_query#L879)):*
> ```
> "https://test.com/path"
> ```

nit: double newline.
___
*[`pkg/util/jsonpath/eval/operation.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OMF51O00XHNe6aa0taa:-OMF51O00XHNe6aa0tab:bkue80f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3366e206d87cdb840bc9f2326c4bf396bfac0ed4/pkg/util/jsonpath/eval/operation.go#L115)):*
> ```Go
> 
> 	regex := op.Right.(jsonpath.Regex)
> 	regex.Regex, err = eval.UnescapePattern(regex.Regex, `\`, false)
> ```

Hm, I think this might be problematic since we're modifying the right operand in place. What will happen if we evaluate the same jsonpath like_regex expression on multiple JSON objects (i.e. multiple rows from the table)? I'd guess that we'd keep unescaping the right side every time, in-place.

Also nit: add inlined comments like `/* escapeToken */` and for `false` too.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I did a quick run of relevant `pg_regress` files, and one difference caught my eye:
```sql
select jsonb_path_query('[null, 1, "abc", "abd", "aBdC", "abdacb", "babc", "adc\nabc", "ab\nadc"]', 'lax $[*] ? (@ like_regex "^ab.*c")');
```
PG returns two rows whereas we return three. At the same time we do match PG on:
```sql
select 'ab\nadc' like '^ab.*c';
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-:-OMF8f_O5ZkX6QGLD4bA:b-js41fb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`pkg/util/jsonpath/eval/operation.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OMF51O00XHNe6aa0taa:-OMF8B5X7hsn1NLZao9T:buzw3xj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3366e206d87cdb840bc9f2326c4bf396bfac0ed4/pkg/util/jsonpath/eval/operation.go#L115)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I think this might be problematic since we're modifying the right operand in place. What will happen if we evaluate the same jsonpath like_regex expression on multiple JSON objects (i.e. multiple rows from the table)? I'd guess that we'd keep unescaping the right side every time, in-place.

Also nit: add inlined comments like `/* escapeToken */` and for `false` too.
</blockquote></details>

Hm, on the second thought, we are operating with the value, not the pointer, so perhaps this will be fine. Actually, your tests should already cover this - we should be invoking `evalRegex` multiple times when we have multiple paths.


<!-- Sent from Reviewable.io -->

normanchenn committed:
jsonpath: add `like_regex` support

This commit add `like_regex` predicate evaluation support. Flags for
`like_regex` are not supported yet.

Epic: None
Release note (sql change): Add `like_regex` predicate evaluation support
for jsonpath queries. Flags for `like_regex` are not supported yet.

normanchenn reviewed (commented):
Should be fixed now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-:-OMJ70Xu1GK31BcQoQac:bql6urd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/util/jsonpath/eval/operation.go` line 115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-OMF51O00XHNe6aa0taa:-OMJ7EGK26-6Je5hooGK:b-3hl0t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3366e206d87cdb840bc9f2326c4bf396bfac0ed4/pkg/util/jsonpath/eval/operation.go#L115)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, on the second thought, we are operating with the value, not the pointer, so perhaps this will be fine. Actually, your tests should already cover this - we should be invoking `evalRegex` multiple times when we have multiple paths.
</blockquote></details>

I changed this to not use `UnescapePattern` anymore, as setting `allowEscapes = true` during scanning and retrieving the json string with `AsText()` seems to do the trick.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 13 of 13 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143240#-:-OMJxMF2A7f5I0-an-7X:b-8ewgme)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @normanchenn)



<!-- Sent from Reviewable.io -->

normanchenn commented:
TFTR!

bors r+
